File: lib/python/src/cabincrew_protocol/artifact.py
from typing import Any, Dict, Optional, Union, List
from pydantic import BaseModel, Field

class Artifact(BaseModel):
    """
    Canonical artifact interface.
    Defined in schemas/draft/artifact.schema.json
    """
    artifact_type: str = Field(..., description="Type of artifact (file, diff, patch, action, message, etc). Free-form and engine-defined.")
    action: str = Field(..., description="Operation to perform with this artifact (create, update, delete, apply, execute, etc). Free-form and engine-defined.")
    target: Optional[str] = Field(None, description="Path, resource, or identifier this artifact applies to. Optional.")
    mime: str = Field(..., description="MIME type describing content.")
    
    # Body can be string, object, array, or null.
    body: Optional[Union[str, Dict[str, Any], List[Any]]] = Field(None, description="Inline content for small artifacts.")
    
    body_file: Optional[str] = Field(None, description="Indicates an external data file within the artifact directory.")
    metadata: Optional[Dict[str, Any]] = Field(None, description="Arbitrary metadata. Optional.")

File: lib/nodejs/src/types/artifact.ts
/**
 * Canonical artifact interface.
 * Defined in schemas/draft/artifact.schema.json
 */
export interface Artifact {
    /**
     * Type of artifact (file, diff, patch, action, message, etc). Free-form and engine-defined.
     */
    artifact_type: string;
    /**
     * Operation to perform with this artifact (create, update, delete, apply, execute, etc). Free-form and engine-defined.
     */
    action: string;
    /**
     * Path, resource, or identifier this artifact applies to. Optional.
     */
    target?: string;
    /**
     * MIME type describing content.
     */
    mime: string;
    /**
     * Inline content for small artifacts.
     * Can be string, object, array, or null.
     */
    body?: string | object | any[] | null;
    /**
     * Indicates an external data file within the artifact directory.
     */
    body_file?: string;
    /**
     * Arbitrary metadata. Optional.
     */
    metadata?: Record<string, any>;
}
